### PR TITLE
random walker: Display a warning when the probability is outsite [0,1] for a given tol

### LINKED ISSUE #2360: Numerical inaccuracies with the random_walker
## Description

The random walker segmentation code is very useful but behaves sometimes very strangely with respect to numerical results.

See following code (sorry this is slow):

result:

Should be between 0 and 1: [0.000000 ; 1.103701]
## Way to reproduce
```
import numpy as np
from skimage.filters import gaussian
from skimage.segmentation import random_walker

np.random.seed(42)
blob3d = np.random.randn(101,101,101) # 3D image of random noise
blob3ds = gaussian(blob3d,sigma=10.) # makes a Gaussian random field
p05=np.percentile(blob3ds,5.) # Normalize the GRF between 0 and 1
p95=np.percentile(blob3ds,95.)
blob3ds = (blob3ds-p05)/(p95-p05)
blob3ds = np.clip(blob3ds,0,1)

markers3d = np.zeros((101,101,101)) 
markers3d[45:55,45:55,45:55] = 1  # central box
markers3d[:,:,0]=markers3d[:,0,:]=markers3d[0,:,:] = 2 # border of the box
markers3d[100,:,:]=markers3d[:,100,:] = markers3d[:,:,100] = 2 # border of the box cont.

grow3d = random_walker(blob3ds, markers3d, beta=5e4, mode='cg_mg', tol=1e-5, return_full_prob=True)

print("Should be between 0 and 1: [%f ; %f]" % (np.min(grow3d[1]), np.max(grow3d[1])))
```

### PR DESCRIPTION
**Pull request by sciunto**
## Description

I fully agree with the diagnosis explained in this https://github.com/scikit-image/scikit-image/issues/2360#issuecomment-326753181

In this PR, I add a new parameter to display a warning when the probability is off.

Closes #2360

## Checklist

- [ ] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [ ] Gallery example in `./doc/examples` (new features only)
- [ ] Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- [ ] Unit tests
- [ ] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)

## For reviewers

- [ ] Check that the PR title is short, concise, and will make sense 1 year
  later.
- [ ] Check that new functions are imported in corresponding `__init__.py`.
- [ ] Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.

### CODE REVIEW

File: skimage/segmentation/random_walker_segmentation.py [hunk 1]
```diff
@@ -484,6 +488,11 @@ def random_walker(data, labels, beta=130, mode='cg_j', tol=1.e-3, copy=True,
     # first at pixel j by anisotropic diffusion.
     X = _solve_linear_system(lap_sparse, B, tol, mode)
 
+    if -X.min() < prob_tol or X.max() - 1 < prob_tol:
```
Comment (rfezzani) [inline skimage/segmentation/random_walker_segmentation.py:491]:
```suggestion
    if X.min() < -prob_tol or X.max() > 1 + prob_tol:
```

File: skimage/segmentation/random_walker_segmentation.py [hunk 2]
```diff
@@ -484,6 +488,11 @@ def random_walker(data, labels, beta=130, mode='cg_j', tol=1.e-3, copy=True,
     # first at pixel j by anisotropic diffusion.
     X = _solve_linear_system(lap_sparse, B, tol, mode)
 
+    if X.min() < -prob_tol or X.max() > 1 + prob_tol:
+        warn('The probability range is outside [0, 1] given the tolerance '
+             '`prob_tol`. Consider decreasing `beta` and/or decreasing '
+             '`tol`'.)
```
Comment (emmanuelle) [inline skimage/segmentation/random_walker_segmentation.py:494]:
```suggestion
             '`tol`.')
```

**Comment by emmanuelle** [comment]
@sciunto I tried to fix the CI problem, please pull from your remote branch :-).

**Comment by emmanuelle** [comment]
Thank you very much @sciunto for addressing an old but relevant issue. Since you added a new attribute could you please also add a new test (probably checking that the warning is issued if `prob_tol` is very low?).

**Comment by sciunto** [comment]
Thanks for the fix and the comment. I'll add some tests and might need to revisit others as well.

**Comment by emmanuelle** [comment]
Now you have to catch the warnings in the tests :-).

**Comment by sciunto** [comment]
Now, it should pass, and proper tests are added. :)

File: skimage/segmentation/tests/test_random_walker.py [hunk 3]
```diff
@@ -430,10 +432,50 @@ def test_isolated_seeds():
...
+    with expected_warnings([NUMPY_MATRIX_WARNING,
+                            'The probability range is outside']):
+        res = random_walker(a, mask, return_full_prob=True)
+    assert res[0, 1, 1] == 1
+    assert res[1, 1, 1] == 0
+
+
+def test_prob_tol():
+    # Turn warnings into errors
+    warnings.simplefilter('error')
```
Comment (hmaarrfk) [inline skimage/segmentation/tests/test_random_walker.py:448]: I think this is good for debugging, but I don't think it should persist through the test suite.

Comment (sciunto) [reply skimage/segmentation/tests/test_random_walker.py:448]: I'm just afraid that otherwise, we miss a failure. What's your suggestion?

Comment (emmanuelle) [reply skimage/segmentation/tests/test_random_walker.py:448]: ah good catch. Anyway uncaught warnings make the test suite fail.

Comment (sciunto) [reply skimage/segmentation/tests/test_random_walker.py:448]: @emmanuelle shall I understand that you want me also to remove this line?

On my machine, the test was not failing without this line. Is there another configuration in the CI that I missed? I remember that warnings were supposed to fail at some point, but I can't remember how this is set.

Comment (sciunto) [reply skimage/segmentation/tests/test_random_walker.py:448]: Found it. The same line is in setup_test()

Comment (jni) [reply skimage/segmentation/tests/test_random_walker.py:448]: @sciunto so are you going to remove it?? =) At which point you will have my approval. =)

Comment (sciunto) [reply skimage/segmentation/tests/test_random_walker.py]: Now that I understand, yes :D

Comment (sciunto) [reply skimage/segmentation/tests/test_random_walker.py]: I just pushed the commit

**Comment by sciunto** [comment]
Who wants to make a second review over here? :)

File: skimage/segmentation/tests/test_random_walker.py [hunk 4]
```diff
@@ -1,4 +1,5 @@
 import numpy as np
+import warnings
```
Comment (rfezzani) [inline skimage/segmentation/tests/test_random_walker.py:2]:
```suggestion
```
I think this is imported but not used

Comment (sciunto) [reply skimage/segmentation/tests/test_random_walker.py:2]: You are correct.

**Pull request merged**